hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3A.java

```diff
@@ -54,4 +55,13 @@ public String toString() {
     sb.append('}');
     return sb.toString();
   }
+
+  /**
+   * Close the file system; the FileContext API doesn't have an explicit close.
+   */
+  @Override
+  protected void finalize() throws Throwable {
```

Comment:
Is it because Hadoop still relies on JDK 8? And cleaner is available from JDK 9? There are also usages of finalize() in other classes like ZlibCompressor. May be we can change all finalize() usages in hadoop when min JDK version moves to 9.

Reply:
one fix would be for FileContext to be closeable and so trigger cleanups without hoping finalize() gets called, especially as GC is triggered by heap use, and i don't know if thread count does that

Reply:
Agreed. May have to change finalize() when JDK version support is upgraded in hadoop.

w.r.t GC causes, it may not look at thread count as a reason for triggering GC.

https://github.com/AdoptOpenJDK/openjdk-jdk8u/blob/master/hotspot/src/share/vm/gc_interface/gcCause.hpp#L39 

https://github.com/AdoptOpenJDK/openjdk-jdk8u/blob/master/hotspot/src/share/vm/gc_interface/gcCause.cpp#L28